src/mender/remoteterminal/remoteterminal.py

```diff
@@ -0,0 +1,217 @@
+
+log = logging.getLogger(__name__)
+
+#mender-connect protocol messeges, taken from:
+#/mendersoftware/go-lib-micro/ws/shell/model.go
+MESSAGE_TYPE_SHELL_COMMAND = 'shell'
+MESSAGE_TYPE_SPAWN_SHELL   = 'new'
+MESSAGE_TYPE_STOP_SHELL    = 'stop'
+
+
+class RemoteTerminal:
+    """ This class serves the RemoteTerminal aka a remote shell feature
+        over the WebSocket. This supposed to be the only instance and
+        serves single connection
+    """
+
+    def __init__(self):
+        log.info("RemoteTerminal initialized")
+
+        self.client = None
+        self.sid = None
+        self.ws_connected = False
+        self.context = None
+        self.session_started = False
+        self.ext_headers = None
+        self.ssl_context = None
+        self.master = None
+        self.slave = None
+        self.shell = None
+
+
+    async def ws_connect(self):
+        '''connects to the backend websocket server.'''
+
+        # from config we receive sth like: "ServerURL": "https://docker.mender.io"
+        # we need replace the protcol and API entry point to achive like this:
+        # "wss://docker.mender.io/api/devices/v1/deviceconnect/connect"
+        uri = self.context.config.ServerURL.replace(
+            "https", "wss") + "/api/devices/v1/deviceconnect/connect"
+        try:
+            self.client = await websockets.connect(
+                uri, ssl=self.ssl_context, extra_headers=self.ext_headers)
+            self.ws_connected = True
+            log.debug(f'connected to: {uri}')
+        except WebSocketException as ws_exception:
+            log.error(f'ws_connect: {ws_exception}')
+
+
+    async def proto_msg_processor(self):
+        '''after having connected to the backend it processes the protocol messages.
+        This function is called in a thread.'''
+
+        await self.ws_connect()
+        if self.client is None:
+            log.debug('Websocket connection failed')
+            return -1
+        log.debug('Websocket connected')
+        try:
+            while True:
+                packed_msg = await self.client.recv()
+                msg: dict = msgpack.unpackb(packed_msg, raw=False)
+                hdr = msg['hdr']
+                if hdr['typ'] == MESSAGE_TYPE_SPAWN_SHELL and not self.session_started:
+                    self.sid = hdr['sid']
+                    self.session_started = True
+                    await self.send_client_status_to_backend(MESSAGE_TYPE_SPAWN_SHELL)
+                    self.open_terminal()
+                    self.start_transmitting_thread()
+                if hdr['typ'] == MESSAGE_TYPE_SHELL_COMMAND:
+                    self.write_command_to_shell(msg)
+                if hdr['typ'] == MESSAGE_TYPE_STOP_SHELL:
+                    self.shell.kill()
+                    self.master = None
+                    self.slave = None
+                    await self.send_client_status_to_backend(MESSAGE_TYPE_STOP_SHELL)
+                    self.session_started = False
```

Comment:
(Not required) Does not handle "resize" message types.

Reply:
According to the SoW we don't support resizing in this version.